Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp

```diff
@@ -109,15 +108,54 @@ void AsyncScrollingCoordinator::handleWheelEventPhase(ScrollingNodeID nodeID, Pl
 
 RefPtr<ScrollingStateNode> AsyncScrollingCoordinator::stateNodeForNodeID(ScrollingNodeID nodeID) const
 {
-    return m_scrollingStateTree->stateNodeForID(nodeID);
+    for (auto& tree : m_scrollingStateTrees.values()) {
+        if (RefPtr scrollingNode = tree->stateNodeForID(nodeID))
+            return scrollingNode;
+    }
+    return nullptr;
 }
 
 RefPtr<ScrollingStateNode> AsyncScrollingCoordinator::stateNodeForScrollableArea(const ScrollableArea& scrollableArea) const
 {
     auto scrollingNodeID = scrollableArea.scrollingNodeID();
     if (!scrollingNodeID)
         return nullptr;
-    return m_scrollingStateTree->stateNodeForID(scrollingNodeID);
+    if (auto* scrollingStateTree = existingScrollingStateTreeForRootFrameID(scrollableArea.rootFrameID()))
+        return scrollingStateTree->stateNodeForID(scrollingNodeID);
+    return nullptr;
+}
+
+ScrollingStateTree& AsyncScrollingCoordinator::ensureScrollingStateTreeForRootFrameID(FrameIdentifier identifier)
+{
+    auto it = m_scrollingStateTrees.find(identifier);
+    if (it != m_scrollingStateTrees.end())
+        return it->value.get();
+
+    ASSERT_WITH_MESSAGE(!page() || page()->settings().siteIsolationEnabled() || m_scrollingStateTrees.isEmpty(), "With site isolation off, we should only ever have one scoling state tree. Get the frame identifier from LocalFrame::rootFrame.");
```

Comment:
This assertion can now be replaced by ASSERT(Frame::isRootFrameIdentifier(identifier)) at the beginning of the function.

Reply:
Fixed.